test/integration/consumer_group_implementation_test.exs

```diff
@@ -0,0 +1,277 @@
+    message
+  end
+
+  def correct_last_message?(nil, _, _), do: false
+  def correct_last_message?([], _, _), do: false
+  def correct_last_message?(message_set, expected_message, expected_offset) do
+    Logger.debug(fn ->
+      "Got message set: #{inspect message_set} " <>
+        "expecting '#{expected_message}' @ offset #{expected_offset}"
+    end)
+    message = List.last(message_set)
+    message.value == expected_message && message.offset == expected_offset
+  end
+
+  def sync_stop(pid) when is_pid(pid) do
+    wait_for(fn ->
+      if Process.alive?(pid) do
+        Process.exit(pid, :normal)
+      end
+      !Process.alive?(pid)
+    end)
+  end
+
+  setup do
+    {:ok, _} = TestPartitioner.start_link
+    {:ok, consumer_group_pid1} = ConsumerGroup.start_link(
+      TestConsumer,
+      @consumer_group_name,
+      [@topic_name],
+      heartbeat_interval: 100,
+      partition_assignment_callback: &TestPartitioner.assign_partitions/2
+    )
+    {:ok, consumer_group_pid2} = ConsumerGroup.start_link(
+      TestConsumer,
+      @consumer_group_name,
+      [@topic_name],
+      heartbeat_interval: 100,
+      partition_assignment_callback: &TestPartitioner.assign_partitions/2
+    )
+
+    # wait for both consumer groups to join
+    wait_for(fn ->
+      ConsumerGroup.active?(consumer_group_pid1) &&
+        ConsumerGroup.active?(consumer_group_pid2)
+    end)
+
+    on_exit fn ->
+      sync_stop(consumer_group_pid1)
+      sync_stop(consumer_group_pid2)
+    end
+
+    {
+      :ok,
+      consumer_group_pid1: consumer_group_pid1,
+      consumer_group_pid2: consumer_group_pid2
+    }
+  end
+
+  test "basic startup, consume, and shutdown test", context do
+    assert TestPartitioner.calls > 0
+
+    generation_id1 = ConsumerGroup.generation_id(context[:consumer_group_pid1])
+    generation_id2 = ConsumerGroup.generation_id(context[:consumer_group_pid2])
+    assert generation_id1 == generation_id2
+
+    member1 = ConsumerGroup.member_id(context[:consumer_group_pid1])
+    member2 = ConsumerGroup.member_id(context[:consumer_group_pid2])
+    assert member1 != member2
+
+    leader1 = ConsumerGroup.leader_id(context[:consumer_group_pid1])
+    leader2 = ConsumerGroup.leader_id(context[:consumer_group_pid2])
+    assert leader1 == leader2
+
+    cond do
+      leader1 == member1 ->
+        assert ConsumerGroup.leader?(context[:consumer_group_pid1])
+        refute ConsumerGroup.leader?(context[:consumer_group_pid2])
+      leader1 == member2 ->
+        refute ConsumerGroup.leader?(context[:consumer_group_pid1])
+        assert ConsumerGroup.leader?(context[:consumer_group_pid2])
+      true ->
+        raise "Neither member is the leader"
+    end
+
+    assignments1 = ConsumerGroup.assignments(context[:consumer_group_pid1])
+    assignments2 = ConsumerGroup.assignments(context[:consumer_group_pid2])
+    assert 2 == length(assignments1)
+    assert 2 == length(assignments2)
+    refute assignments1 == assignments2
```

Comment:
maybe assert that none of the assignments in `assignments1` are in `assignments2`